### DISCUSSION TITLE
Status of the project

### DISCUSSION THREAD

**Discussion by krisc-informatica** [krisc-informatica/arduino-smart-bike-trainer#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by hchris1** [comment]
Hi @krisc-informatica,

I own a Tacx Cycleforce Basic myself and this project seems really interesting. Maybe you could answer a few questions for me:
- Could you eleborate on how the project advanced? Is it WIP or do you have it working already?
- Are there plans for the electronics involved in this?

Thank you for your help

**Reply by krisc-informatica** [reply]
Hello @hchris1,

Currently, the project is WIP but it has evolved quite a lot already. The github project does not yet contain the latest tests that I performed.

What I have done so far:
- synchronization for brake PWM to the 23VAC signal pin. I have used a diode and voltage divider to reduce the voltage to 3.3V with pulses (high/low). I have measured the pwm periods that are used from the stearing console and can generate them in sync with the signal.
- the hardware is finished in prototype (break out board). I have made a pcb layout here: https://oshwlab.com/kris.cardinaels/arduino-smart-bike-trainer

**Reply by hchris1** [reply]
Thank you for your answer and the update!

That sounds really promising. Did you already find something out on how to get the current wattage? That info would probably be the most important one. 

I have some experience with Arduino and electronics myself so whenever my time allows I would be happy contribute.

**Reply by krisc-informatica** [reply]
For the power/resistance I was thinking to check out https://support.strava.com/hc/en-us/articles/216917107-How-Strava-Calculates-Power and https://www.omnicalculator.com/sports/cycling-wattage#what-is-the-cycling-wattage which are more or less the same.

These formules provide a wattage needed for a ride. I would have to use a function to map it to resistance in the brake.

The Tacx flow has a fixed resistance for the different inclinations (-4..9 %), not bound to the weight or rolling resistance etc. But those values for the PWM could help in setting the brake.

**Reply by alexphredorg** [reply]
@krisc-informatica Is the status more or less the same since Jan 22?  Or are there some updates that you haven't pushed yet?

**Comment by krisc-informatica** [comment]
A new test this week showed good progress!

The actual power calculation should me optimized, but power sent from trainer to Zwift is acceptable, although about 50 Watt off...
The brake control works well for Zwift percentages of 6 and up. Below, the control seems somewhat unstable. I have to control it with the oscilloscope to compare the trainer pulses with the tacx pulses sent from the steer hub.

But, I could ride the Mont Ventoux and it felt hard on some parts of the track :-)

**Reply by StFalagar** [reply]
Congratulations... it's quite interesting that it works well for 6 and up... my experience with the Fortius and Flow i.e. motorbrake and eddy current brake is that they work well up to 5%.  Especially at low cadence going over 5% doesn't work well... the Eddy Current brake has a band of 13 resistance levels and above 5% it is difficult to keep the cadence high enough to apply the power within the band.  The motorbrake has an awesome range but becomes jittery at high power and low cadence... i.e at about 6 percent at low cadence.

**Comment by luc-h77** [comment]
Meanwhile I've managed to make my Flow 1680 a complete smart trainer, while zwifting I don't feel much of a difference to the Kickr. 
My setup is a bit different, I'm running a nodejs program on my laptop which is taking care of the bluetooth part, I'll share that soon. However the power calculation in the Arduino could also be relevant for you.
I've done quite a bit of power testing against my Power2Max Crank and now I'm getting a really good output. Today's 100k ride on Zwift came out with an average of 201 off the crank versus 203 off the trainer.
Important is the a spindown test. I'm riding a conti indoor tyre with 7.5 bar. Accelerate to 50km/h in the lowest gear and the spindown should take 14 seconds.
By comparison, if the spindown takes 19 seconds I have to push 30-60 watts less.

Also noticed that the first 15 minutes the trainer is underreporting by up to 15%. This seems to be due to the cold tyre. So I'm thinking of adding a decreasing factor during the initial 15 minutes of a ride (1.15 at minute 0 down to 1.01 at minute 14).
Attached my Arduino file and the data of today's ride.
[Tacx Controller Calibrated.txt](https://github.com/krisc-informatica/arduino-smart-bike-trainer/files/7109713/Tacx.Controller.Calibrated.txt)
[Zwift TBR 101 20210904.xlsx](https://github.com/krisc-informatica/arduino-smart-bike-trainer/files/7109714/Zwift.TBR.101.20210904.xlsx)
